### PR TITLE
feat(optimizer): support common sub-expression extraction

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Support `CommonSubExprExtractRule` which would extract common sub-expression into a new project as an input of the original project.
- let `ProjectMergeRule` detect whether the inner project is generated by `CommonSubExprExtractRule`, if so, we don't merge these projects.
- Remove project merge logic from `to_stream` and `to_batch` and use `StreamProjectMergeRule` and `BatchProjectMergeRule` instead.
- Resolve https://github.com/risingwavelabs/risingwave/issues/10812

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/frontend/src/optimizer/plan_expr_visitor/expr_counter.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    fn merge(_: (), _: ()) {}
+
+    fn visit_expr(&mut self, expr: &ExprImpl) {
+        // Considering this sql, `In` expression needs to ensure its in-clauses to be const.
+        // If we extract it into a common sub-expression (finally be a `InputRef`) which will
+        // violate this assumption, so ban this case. SELECT x,
+        //        tand(x) IN ('-Infinity'::float8,-1,0,1,'Infinity'::float8) AS tand_exact,
+        //        cotd(x) IN ('-Infinity'::float8,-1,0,1,'Infinity'::float8) AS cotd_exact
+        // FROM (VALUES (0), (45), (90), (135), (180),(225), (270), (315), (360)) AS t(x);
+        if expr.is_const() {
```
Comment (BugenZhao) [inline src/frontend/src/optimizer/plan_expr_visitor/expr_counter.rs:36]: `is_const` will perform a traversal on the given expression. Not sure if it hurts performance. 🥵

File: src/frontend/planner_test/tests/testdata/output/join.yaml [hunk 2]
```diff
@@ -506,27 +506,29 @@
...
-            └─BatchHashJoin { type: FullOuter, predicate: a.x = b.x, output: all }
-              ├─BatchExchange { order: [], dist: HashShard(a.x) }
-              │ └─BatchScan { table: a, columns: [a.x], distribution: SomeShard }
-              └─BatchExchange { order: [], dist: HashShard(b.x) }
-                └─BatchScan { table: b, columns: [b.x], distribution: SomeShard }
+    BatchProject { exprs: [$expr3, $expr2] }
+    └─BatchExchange { order: [$expr2 ASC], dist: Single }
+      └─BatchProject { exprs: [(2:Int32 * $expr1) as $expr3, $expr2, $expr2] }
+        └─BatchSort { order: [$expr2 ASC] }
+          └─BatchProject { exprs: [a.x, b.x, Coalesce(a.x, b.x) as $expr1, (Coalesce(a.x, b.x) + Coalesce(a.x, b.x)) as $expr2] }
```
Comment (BugenZhao) [inline src/frontend/planner_test/tests/testdata/output/join.yaml:513]: Seems `Coalesce(a.x, b.x)` is not extracted here?

Comment (chenzl25) [reply src/frontend/planner_test/tests/testdata/output/join.yaml:513]: Good Catch! We should apply this rule in top down order.

File: src/frontend/src/optimizer/plan_expr_visitor/expr_counter.rs (hunk 1 above)
Comment (chenzl25) [reply src/frontend/src/optimizer/plan_expr_visitor/expr_counter.rs:36]: Maybe we can enable this rule only for streaming queries.

**Review by st1page — APPROVED** [review]

**Pull request merged**